File: poc/backend/app/schemas/case.py
from datetime import date, datetime
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from .common import PaginationQuery


class CaseListQuery(PaginationQuery):
    status: str | None = None
    pool_type: str | None = None
    assigned_to: int | None = None
    keyword: str | None = None


class CaseResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    tenant_id: int
    project_id: int | None
    owner_id: int
    assigned_to: int | None
    pool_type: str
    stage: str
    amount_owed: Decimal | None
    months_overdue: int | None
    priority_score: int
    last_contact_at: datetime | None
    monthly_contact_count: int
    status: str
    created_at: datetime
    updated_at: datetime


class CaseAssignRequest(BaseModel):
    case_ids: list[int] = Field(..., min_length=1, max_length=500)
    assign_to: int


class CaseImportRow(BaseModel):
    name: str = Field(..., min_length=1)
    phone: str = Field(..., pattern=r"^1[3-9]\d{9}$")
    building: str | None = None
    room: str | None = None
    amount_owed: Decimal | None = None
    months_overdue: int | None = None
    # v1.6.3 — 账单字段（导入时直接录入，不再按月推算明细）
    bill_period_start: date | None = None
    bill_period_end: date | None = None
    principal_amount: Decimal | None = None  # 物业费
    late_fee_amount: Decimal | None = None  # 违约金 / 滞纳金
    arrears_reason: str | None = None  # 经济困难 / 服务质量异议 / 房屋空置 / 租客拖欠 / 其他
    notes: str | None = None  # 欠费情况说明（拒缴/暂时困难/失联等）


class OwnerInfo(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    phone: str | None = None  # decrypted, only for agent_internal
    phone_masked: str
    building: str | None
    room: str | None
    do_not_call: bool


class CaseWithOwnerResponse(BaseModel):
    id: int
    tenant_id: int
    project_id: int | None
    project_name: str | None = None
    owner: OwnerInfo
    assigned_to: int | None
    pool_type: str
    stage: str
    amount_owed: Decimal | None
    months_overdue: int | None
    priority_score: int
    last_contact_at: datetime | None
    monthly_contact_count: int
    status: str
    notes: str | None = None
    created_at: datetime
    updated_at: datetime
    # v1.5.5 — 项目所属服务商（无项目 / 物业自办时为 None）
    provider_id: int | None = None
    provider_name: str | None = None


class CaseImportRequest(BaseModel):
    project_id: int | None = None
    rows: list[CaseImportRow] = Field(..., min_length=1, max_length=500)


class CaseImportResponse(BaseModel):
    imported: int
    skipped: int
    errors: list[str]


class CaseStageUpdate(BaseModel):
    stage: Literal["new", "in_progress", "promised", "paid", "escalated", "closed"]
    note: str | None = None  # v1.6.6 — 阶段变更跟进备注（写入 audit log）


class CaseAssignResponse(BaseModel):
    updated_count: int


# ── Sprint 3b: case detail with call timeline ──────────────────


class CaseCallItem(BaseModel):
    id: int
    started_at: datetime | None
    duration_sec: int | None
    status: str
    transcript_preview: str | None
    result_tag: str | None
    confidence: float | None
    agent_name: str | None
    recording_url: str | None = None  # v1.6.7 — E5 inline 录音播放


class TimelineEvent(BaseModel):
    type: str
    ts: datetime
    actor: str | None
    note: str | None
    # v1.6.9 — 关联实体 ID，让前端能跳到对应详情页（工单/法务订单/通话）
    target_id: int | None = None
    # 'workorder' / 'legal_order' / 'legal_case' / 'call' / 'audit'
    target_type: str | None = None


class CaseProjectInfo(BaseModel):
    """v1.6.3 — 案件详情中嵌入项目基本信息（合同 + 收费），无需另起接口。"""

    name: str
    charge_rate_text: str | None = None
    charge_period: str | None = None
    contract_type: str | None = None
    contract_start_date: date | None = None
    contract_end_date: date | None = None
    contract_attachment_key: str | None = None
    contract_attachment_filename: str | None = None
    charge_notes: str | None = None


class CaseDetailResponse(BaseModel):
    id: int
    tenant_id: int
    project_id: int | None
    project_name: str | None = None
    project_info: CaseProjectInfo | None = None  # v1.6.3
    owner: OwnerInfo
    assigned_to: int | None
    assigned_role: str | None = (
        None  # v1.4 — 协作来源 badge：agent_internal / agent_external / null
    )
    pool_type: str
    stage: str
    amount_owed: Decimal | None
    months_overdue: int | None
    # v1.6.3 — 账单字段（导入时录入，不再按月推算）
    bill_period_start: date | None = None
    bill_period_end: date | None = None
    principal_amount: Decimal | None = None
    late_fee_amount: Decimal | None = None
    arrears_reason: str | None = None
    priority_score: int
    last_contact_at: datetime | None
    monthly_contact_count: int
    # v1.8.0 — 业主画像 3 统计卡片：联系次数复用 monthly_contact_count
    promise_count: int = 0
    workorder_count: int = 0
    status: str
    created_at: datetime
    updated_at: datetime
    calls: list[CaseCallItem]
    timeline_events: list[TimelineEvent]
    # v1.5 — 服务团队（电话团队 = 项目签约的服务商；法务团队 = 转化订单已撮合的律所）
    calling_provider_id: int | None = None
    calling_provider_name: str | None = None
    legal_law_firm_name: str | None = None
    legal_lawyer_name: str | None = None
    legal_order_status: str | None = None

File: frontend/src/types/case.ts
export interface CaseCallItem {
  id: number;
  started_at: string | null;
  duration_sec: number | null;
  status: string;
  transcript_preview: string | null;
  result_tag: string | null;
  confidence: number | null;
  agent_name: string | null;
  recording_url?: string | null;  // v1.6.7 — E5 inline 录音
}

export interface TimelineEvent {
  type: string;
  ts: string;
  actor: string | null;
  note: string | null;
  // v1.6.9 — 关联实体 ID + 类型，让前端能跳到对应详情页
  target_id?: number | null;
  target_type?: "workorder" | "legal_order" | "legal_case" | "call" | "audit" | null;
}

export interface OwnerInfo {
  id: number;
  name: string;
  phone: string | null;
  phone_masked: string;
  building: string | null;
  room: string | null;
  do_not_call: boolean;
}

export interface CaseProjectInfo {
  name: string;
  charge_rate_text: string | null;
  charge_period: string | null;
  contract_type: string | null;
  contract_start_date: string | null;
  contract_end_date: string | null;
  contract_attachment_key: string | null;
  contract_attachment_filename: string | null;
  charge_notes: string | null;
}

export interface CaseDetailResponse {
  id: number;
  tenant_id: number;
  project_id: number | null;
  owner: OwnerInfo;
  assigned_to: number | null;
  // v1.0.0 — 案件详情显示催收员姓名(物业 + 服务商 admin 详情页都用)
  assigned_to_name?: string | null;
  pool_type: string;
  stage: string;
  amount_owed: string | null;
  months_overdue: number | null;
  // v1.6.3 — 账单字段（导入时录入，不再按月推算）
  bill_period_start?: string | null;
  bill_period_end?: string | null;
  principal_amount?: string | null;
  late_fee_amount?: string | null;
  arrears_reason?: string | null;
  priority_score: number;
  last_contact_at: string | null;
  monthly_contact_count: number;
  // v1.8.0 — 业主画像 3 统计卡片
  promise_count?: number;
  workorder_count?: number;
  status: string;
  created_at: string;
  updated_at: string;
  calls: CaseCallItem[];
  timeline_events: TimelineEvent[];
  project_name?: string | null;
  project_info?: CaseProjectInfo | null;  // v1.6.3
  assigned_role?: string | null;
  // v1.5 — 服务团队
  calling_provider_id?: number | null;
  calling_provider_name?: string | null;
  legal_law_firm_name?: string | null;
  legal_lawyer_name?: string | null;
  legal_order_status?: string | null;
  // v0.6.0 — 案件下「等审批」的法务转化申请;null=无申请。
  // 用于督导端「移交法务 / 审批转法务」按钮条件渲染。
  pending_legal_conversion_request_id?: number | null;
}
